configs/provider_requirements.go

```diff
@@ -58,14 +57,15 @@ func decodeRequiredProvidersBlock(block *hcl.Block) ([]*RequiredProvider, hcl.Di
 						Detail:   "This string does not use correct version constraint syntax.",
 						Subject:  attr.Expr.Range().Ptr(),
 					})
-					reqs = append(reqs, &RequiredProvider{Name: name})
-					return reqs, diags
+				} else {
+					vc.Required = constraints
+					ret.Requirement = vc
 				}
-				vc.Required = constraints
-				reqs = append(reqs, &RequiredProvider{Name: name, Requirement: vc})
 			}
-			// No version
-			reqs = append(reqs, &RequiredProvider{Name: name})
+			if expr.Type().HasAttribute("source") {
+				ret.Source = expr.GetAttr("source").AsString()
+			}
+			reqs = append(reqs, ret)
```

Comment:
I might be misreading this, but it looks like previously we could add two entries to the required provider list, but now we always add one.

(If the requirement had a valid version constraint, we'd add a `RequiredProvider` with a version, and also a versionless `RequiredProvider`. Now we will only add a single `RequiredProvider`, which may have version constraints if they're present and valid.)

Was the previous behaviour just an oversight? I'm not familiar with the consumer(s) of this data structure, but at first glance it looks like all the duplicate provider requirements had their version constraints merged anyway, so the versionless requirement didn't seem to serve any purpose.

Reply:
The change to the flow here is odd. In the refactored funtion, it  will return a versionless constraint if there is a version error: the versionless return (`ret`) is set up at the top of the `case` statement, but now instead of returning immediately if there is a version error, this will continue and parse the source attribute, and only after both of those attributes are parsed will it return. 

I will add some more test cases that cover invalid version (and source) attributes - this function was untested before now and it's been a challenge to refactor and test at the same time 🙃